File: src/stripe_link/domain/slugs.py
"""Slug rules for the whole application — one implementation, not one per entity.

Offers were the first thing to need slugs; products, services and anything else that earns a URL will
need exactly the same rules. Keeping them here means a new entity type inherits the rules instead of
re-deriving them slightly differently, which is how "workout-bundle" and "dietary-supplement-bundle"
came to describe the same offer.

The JS mirror is dashboard/src/composables/slugs.js. The two are held together by
tests/fixtures/slug_cases.json, which BOTH test suites run — parity by shared fixtures, because an
algorithm cannot be shared across the runtimes the way composition_rules.json is.
"""

import re

FALLBACK_SLUG = "offer"

# Dropped from generated slugs: they add length and no search value.
STOP_WORDS = frozenset({
    "the", "a", "an", "and", "or", "for", "of", "with", "to", "in", "on", "at", "by", "from",
    "your", "you", "our", "my", "this", "that", "is", "are", "plus",
})


def sanitize_slug(value, fallback: str = FALLBACK_SLUG) -> str:
    """URL-safe slug: lowercase, non-alphanumerics collapsed to single hyphens, trimmed.

    Deliberately ASCII-only: a non-Latin name degrades to its ASCII remnant rather than emitting
    percent-encoded bytes, which are unreadable in a URL bar and worthless as a search signal.
    """
    slug = re.sub(r"[^a-z0-9]+", "-", str(value or "").strip().lower()).strip("-")
    return slug or fallback


def slug_tokens(text, limit: int = 0) -> list[str]:
    """Meaningful lowercase words from free text, stop-words removed, optionally capped."""
    words = [w for w in re.split(r"[^a-z0-9]+", str(text or "").lower()) if w and w not in STOP_WORDS]
    return words[:limit] if limit else words


def unique_slug(desired, taken, fallback: str = FALLBACK_SLUG) -> str:
    """Sanitize `desired` and make it unique against `taken`, appending -2, -3, … on collision.

    `taken` is the set of slugs already in use, with the CALLER's own current slug removed — an entity
    being re-saved must not collide with itself and renumber its own live URL on every save.
    """
    base = sanitize_slug(desired, fallback)
    taken = {str(s) for s in (taken or ())}
    if base not in taken:
        return base
    candidate, suffix = base, 2
    while candidate in taken:
        candidate = f"{base}-{suffix}"
        suffix += 1
    return candidate

File: dashboard/src/composables/slugs.js
// Slug rules — the JS MIRROR of src/stripe_link/domain/slugs.py.
//
// The server is always authoritative: it alone sees every existing slug at write time, and another tab can
// claim one between a preview and a save. This mirror exists so the builder can show the tenant the slug
// they will actually get, instead of one the server silently renumbers on save.
//
// The two implementations are held together by tests/fixtures/slug_cases.json, which BOTH suites run. An
// algorithm can't be shared across runtimes the way composition_rules.json is, so the FIXTURES are shared.
// Change a rule here and you must change slugs.py, or the parity test fails.

export const FALLBACK_SLUG = "offer";

export const STOP_WORDS = new Set([
  "the", "a", "an", "and", "or", "for", "of", "with", "to", "in", "on", "at", "by", "from",
  "your", "you", "our", "my", "this", "that", "is", "are", "plus",
]);

// ASCII-only on purpose: a non-Latin name degrades to its ASCII remnant rather than emitting
// percent-encoded bytes, which are unreadable in a URL bar and worthless as a search signal.
export function sanitizeSlug(value, fallback = FALLBACK_SLUG) {
  const slug = String(value ?? "").trim().toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "");
  return slug || fallback;
}

export function slugTokens(text, limit = 0) {
  const words = String(text ?? "").toLowerCase().split(/[^a-z0-9]+/).filter((w) => w && !STOP_WORDS.has(w));
  return limit ? words.slice(0, limit) : words;
}

// `taken` must already have the caller's OWN current slug removed — an entity being re-saved must not
// collide with itself and renumber its own live URL.
export function uniqueSlug(desired, taken, fallback = FALLBACK_SLUG) {
  const base = sanitizeSlug(desired, fallback);
  const used = taken instanceof Set ? taken : new Set(Array.from(taken || []).map(String));
  if (!used.has(base)) return base;
  let candidate = base;
  let suffix = 2;
  while (used.has(candidate)) {
    candidate = `${base}-${suffix}`;
    suffix += 1;
  }
  return candidate;
}
